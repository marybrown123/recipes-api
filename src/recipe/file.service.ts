import { ForbiddenException, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';

@Injectable()
export class FileService {
  private s3Client = new S3Client({
    region: this.configService.getOrThrow('AWS_S3_REGION'),
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    },
  });

  constructor(private configService: ConfigService) {}

  async uploadFileToS3(fileName: string, file: Buffer) {
    const params = {
      Bucket: process.env.S3_BUCKET_NAME,
      Key: fileName,
      Body: file,
    };
    const command = new PutObjectCommand(params);
    try {
      await this.s3Client.send(command);
    } catch (error) {
      throw new ForbiddenException();
    }

    return `https://${params.Bucket}.s3.${process.env.AWS_S3_REGION}.amazonaws.com/${params.Key}`;
  }
}
