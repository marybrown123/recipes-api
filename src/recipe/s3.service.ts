import {
  DeleteObjectCommand,
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { randomUUID } from 'crypto';

@Injectable()
export class S3Service {
  private s3Client = new S3Client({
    region: this.configService.getOrThrow('AWS_S3_REGION'),
    credentials: {
      accessKeyId: this.configService.getOrThrow('AWS_ACCESS_KEY_ID'),
      secretAccessKey: this.configService.getOrThrow('AWS_SECRET_ACCESS_KEY'),
    },
  });

  constructor(private configService: ConfigService) {}

  generateS3Key(fileName: string): string {
    return `${randomUUID()}-${fileName}`;
  }

  async generatePresignedUrl(bucket: string, key: string): Promise<string> {
    const params = {
      Bucket: bucket,
      Key: key,
    };

    return getSignedUrl(this.s3Client, new GetObjectCommand(params), {
      expiresIn: 60,
    });
  }

  async uploadFileToS3(
    bucket: string,
    key: string,
    file: Buffer,
  ): Promise<void> {
    const params = {
      Bucket: bucket,
      Key: key,
      Body: file,
    };
    const command = new PutObjectCommand(params);
    await this.s3Client.send(command);
  }

  async deleteFileFromS3(bucket: string, key: string): Promise<void> {
    const params = {
      Bucket: bucket,
      Key: key,
    };
    const command = new DeleteObjectCommand(params);
    await this.s3Client.send(command);
  }
}
