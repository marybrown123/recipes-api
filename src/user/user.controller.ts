import { Body, Controller, Post } from '@nestjs/common';
import * as bcrypt from 'bcrypt';
import { UserService } from './user.service';
import { CreateUserDTO } from './DTOs/create-user.DTO';
@Controller('/user')
export class UserController {
  constructor(private readonly usersService: UserService) {}

  @Post('/signup')
  async createUser(@Body() user: CreateUserDTO) {
    const hashedPassword = await bcrypt.hash(user.password, 10);
    const userForDb = {
      name: user.name,
      password: hashedPassword,
    };
    const createdUser = await this.usersService.createUser(userForDb);
    return createdUser;
  }
}
