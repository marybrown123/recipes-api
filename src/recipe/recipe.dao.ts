import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { RecipeResponse } from '../recipe/responses/recipe.response';
import { CreateRecipeDTO } from 'src/recipe/DTOs/create-recipe.dto';
import { UpdateRecipeDTO } from 'src/recipe/DTOs/update-recipe.dto';
import { FindAllRecipesDTO } from 'src/recipe/DTOs/find-all-recipes-query';

@Injectable()
export class RecipeDAO {
  constructor(private prismaService: PrismaService) {}
  async createRecipe(
    recipe: CreateRecipeDTO,
    authorId: number,
  ): Promise<RecipeResponse> {
    const newRecipe = await this.prismaService.recipe.create({
      data: {
        ...recipe,
        author: {
          connect: {
            id: authorId,
          },
        },
        preparing: { create: recipe.preparing },
        ingredients: { create: recipe.ingredients },
      },
      include: {
        preparing: true,
        ingredients: true,
      },
    });
    return new RecipeResponse(newRecipe);
  }

  async updateRecipe(
    recipeId: number,
    newRecipe: UpdateRecipeDTO,
  ): Promise<RecipeResponse> {
    if (newRecipe.preparing) {
      await this.prismaService.recipePreparationSteps.deleteMany({
        where: {
          recipeId,
        },
      });
    }

    if (newRecipe.ingredients) {
      await this.prismaService.recipeIngredients.deleteMany({
        where: {
          recipeId,
        },
      });
    }
    const updatedRecipe = await this.prismaService.recipe.update({
      where: { id: recipeId },
      data: {
        ...newRecipe,
        preparing: { create: newRecipe.preparing },
        ingredients: { create: newRecipe.ingredients },
      },
      include: {
        preparing: true,
        ingredients: true,
      },
    });

    return new RecipeResponse(updatedRecipe);
  }

  async findRecipeById(recipeId: number): Promise<RecipeResponse> {
    const recipeFromDb = await this.prismaService.recipe.findFirst({
      where: {
        id: recipeId,
      },
      include: {
        preparing: true,
        ingredients: true,
      },
    });

    return new RecipeResponse(recipeFromDb);
  }

  async findAllRecipes(query: FindAllRecipesDTO): Promise<RecipeResponse[]> {
    const whereCodition = query.name ? { name: { contains: query.name } } : {};
    const recipesFromDb = await this.prismaService.recipe.findMany({
      where: whereCodition,
      skip: (query.page - 1) * query.limit,
      take: query.limit,
      include: {
        preparing: true,
        ingredients: true,
      },
    });
    return recipesFromDb.map((recipe) => {
      return new RecipeResponse(recipe);
    });
  }
}
