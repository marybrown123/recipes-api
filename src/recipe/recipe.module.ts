import { Module } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { RecipeService } from './recipe.service';
import { RecipeController } from './recipe.controller';
import { FindRecipeByIdHandler } from './queries/handlers/findRecipeById.handler';
import { CqrsModule } from '@nestjs/cqrs';
import { RecipeDAO } from 'src/recipe/recipe.dao';
import { FindAllRecipesHandler } from 'src/recipe/queries/handlers/findAllRecipes.handler';
import { CreateRecipeHandler } from 'src/recipe/commands/handlers/createRecipe.handler';
import { UpdateRecipeHandler } from 'src/recipe/commands/handlers/updateRecipe.handler';

export const QueryHandlers = [FindRecipeByIdHandler, FindAllRecipesHandler];
export const CommandHandlers = [CreateRecipeHandler, UpdateRecipeHandler];

@Module({
  imports: [CqrsModule],
  providers: [
    RecipeService,
    ...QueryHandlers,
    ...CommandHandlers,
    PrismaService,
    RecipeDAO,
  ],
  exports: [RecipeService],
  controllers: [RecipeController],
})
export class RecipeModule {}
