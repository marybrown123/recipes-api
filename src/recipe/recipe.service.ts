import { Inject, Injectable } from '@nestjs/common';
import { CreateRecipeDTO } from './DTOs/create-recipe.dto';
import { RecipeResponse } from './responses/recipe.response';
import { UpdateRecipeDTO } from './DTOs/update-recipe.dto';
import { FindAllRecipesDTO } from './DTOs/find-all-recipes-query';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { FindRecipeByIdQuery } from './queries/impl/findRecipeById.query';
import { CreateRecipeCommand } from '../recipe/commands/impl/createRecipe.command';
import { UpdateRecipeCommand } from '../recipe/commands/impl/updateRecipe.command';
import { FindAllRecipesQuery } from '../recipe/queries/impl/findAllRecipes.query';
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Cache } from 'cache-manager';

@Injectable()
export class RecipeService {
  constructor(
    private queryBus: QueryBus,
    private commandBus: CommandBus,
    @Inject(CACHE_MANAGER) private cacheService: Cache,
  ) {}

  async createRecipe(
    recipe: CreateRecipeDTO,
    authorId: number,
  ): Promise<RecipeResponse> {
    return this.commandBus.execute(new CreateRecipeCommand(recipe, authorId));
  }

  async updateRecipe(
    recipeId: number,
    newRecipe: UpdateRecipeDTO,
  ): Promise<RecipeResponse> {
    return this.commandBus.execute(
      new UpdateRecipeCommand(recipeId, newRecipe),
    );
  }

  async findRecipeById(recipeId: number): Promise<RecipeResponse> {
    const cachedData = await this.cacheService.get<RecipeResponse>(
      `/recipe/${recipeId}`,
    );
    if (cachedData) {
      return cachedData;
    } else {
      const recipeFromDb = await this.queryBus.execute(
        new FindRecipeByIdQuery(recipeId),
      );
      await this.cacheService.set(`/recipe/${recipeId}`, recipeFromDb);
      return recipeFromDb;
    }
  }

  async findAllRecipes(query: FindAllRecipesDTO): Promise<RecipeResponse[]> {
    return this.queryBus.execute(new FindAllRecipesQuery(query));
  }
}
