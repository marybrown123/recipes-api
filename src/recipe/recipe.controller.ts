import {
  Body,
  Controller,
  Get,
  Param,
  Patch,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';
import { CreateRecipeDTO } from './DTOs/create-recipe.dto';
import { RecipeService } from './recipe.service';
import { AuthGuard } from '@nestjs/passport';
import { CurrentUser } from 'src/common/decorators/user.decorator';
import { UpdateRecipeDTO } from './DTOs/update-recipe.dto';
import { IsUserAuthorGuard } from 'src/user/guards/is-user-author.guard';
import { IsAdminGuard } from 'src/user/guards/is-admin.guard';
import {
  ApiCreatedResponse,
  ApiForbiddenResponse,
  ApiOperation,
  ApiParam,
  ApiResponse,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { RecipeResponse } from './responses/recipe.response';

@Controller('/recipe')
export class RecipeController {
  constructor(private readonly recipeService: RecipeService) {}

  @Post()
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ summary: 'Create a new recipe' })
  @ApiCreatedResponse({ type: RecipeResponse })
  @ApiUnauthorizedResponse({ description: 'Not logged in' })
  async createRecipe(@Body() recipe: CreateRecipeDTO, @CurrentUser() user) {
    return await this.recipeService.createRecipe(recipe, user.id);
  }

  @Patch('/:id')
  @UseGuards(AuthGuard('jwt'), IsUserAuthorGuard)
  @ApiOperation({ summary: 'Update a recipe' })
  @ApiCreatedResponse({ type: RecipeResponse })
  @ApiUnauthorizedResponse({ description: 'Not logged in' })
  @ApiForbiddenResponse({
    description: 'User does not own this recipe',
  })
  @ApiParam({ name: 'id', required: true })
  async updateRecipe(
    @Body() newRecipe: UpdateRecipeDTO,
    @Param('id') recipeId: number,
  ) {
    return await this.recipeService.updateRecipe(Number(recipeId), newRecipe);
  }

  @Get('/byId/:id')
  @ApiOperation({ summary: 'Get one recipe by id' })
  @ApiResponse({ type: RecipeResponse })
  @ApiParam({ name: 'id', required: true })
  async getOneRecipe(@Param('id') recipeId: number) {
    return await this.recipeService.findOneRecipe(Number(recipeId));
  }

  @Get()
  @UseGuards(AuthGuard('jwt'), IsAdminGuard)
  @ApiOperation({ summary: 'Get all recipes' })
  @ApiResponse({ type: [RecipeResponse] })
  @ApiUnauthorizedResponse({ description: 'Not logged in' })
  @ApiForbiddenResponse({
    description: 'User is not an admin',
  })
  async getAllRecipes(
    @Query('page') page: number,
    @Query('limit') limit: number,
  ) {
    return await this.recipeService.findAllRecipes(Number(limit), Number(page));
  }

  @Get('/byName')
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ summary: 'Get a recipe by name' })
  @ApiResponse({ type: [RecipeResponse] })
  @ApiUnauthorizedResponse({ description: 'Not logged in' })
  async getRecipeByName(@Query('query') query: string) {
    return await this.recipeService.findRecipeByName(query);
  }
}
