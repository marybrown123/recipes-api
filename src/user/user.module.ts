import { Module } from '@nestjs/common';
import { UserService } from './user.service';
import { UserController } from './user.controller';
import { PrismaService } from '../prisma/prisma.service';
import { MailModule } from '../mail/mail.module';
import { TokenModule } from '../token/token.module';
import { WebhookModule } from '../webhook/webhook.module';

@Module({
  imports: [MailModule, TokenModule, WebhookModule],
  providers: [UserService, PrismaService],
  exports: [UserService],
  controllers: [UserController],
})
export class UserModule {}
