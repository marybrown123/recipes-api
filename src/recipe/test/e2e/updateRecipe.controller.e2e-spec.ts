import {
  CanActivate,
  HttpStatus,
  INestApplication,
  ValidationPipe,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import { RecipeService } from '../../../recipe/recipe.service';
import { RecipeServiceMock } from '../../../recipe/test/mocks/recipe.service.mock';
import { UpdateRecipeDTO } from '../../../recipe/DTOs/update-recipe.dto';
import { IsUserAuthorGuard } from '../../../user/guards/is-user-author.guard';
import { UserService } from '../../../user/user.service';
import { Role, User } from '@prisma/client';
import { PrismaService } from '../../../prisma/prisma.service';
import { AppModule } from '../../../app.module';

describe('Recipe Controller - Update', () => {
  let app: INestApplication;
  let jwtService: JwtService;
  let userService: UserService;
  let accessToken: string;
  let prismaService: PrismaService;
  let testUser: User;
  const fakeGuard: CanActivate = { canActivate: jest.fn(() => true) };
  const correctPayload: UpdateRecipeDTO = {
    name: 'testName',
    description: 'testDescription',
    image: { name: 'testImageName', key: 'testImageKey' },
    preparing: [
      {
        step: 'testStep',
        order: 1,
      },
    ],
    ingredients: [
      {
        name: 'testName',
        amount: 'testAmount',
      },
    ],
  };
  const validationTestsStructure = [
    {
      property: 'name',
      value: 1,
    },
    {
      property: 'description',
      value: 1,
    },
    {
      property: 'image',
      value: 1,
    },
    {
      property: 'preparing',
      value: 1,
    },
    {
      property: 'ingredients',
      value: 1,
    },
  ];
  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    })
      .overrideProvider(RecipeService)
      .useClass(RecipeServiceMock)
      .overrideGuard(IsUserAuthorGuard)
      .useValue(fakeGuard)
      .compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe({ transform: true }));
    await app.init();
    jwtService = moduleFixture.get<JwtService>(JwtService);
    userService = moduleFixture.get<UserService>(UserService);
    prismaService = moduleFixture.get<PrismaService>(PrismaService);

    testUser = await userService.generateAccount(
      process.env.TEST_NAME,
      process.env.TEST_PASSWORD,
      Role.USER,
    );
    accessToken = jwtService.sign({
      name: testUser.name,
      sub: testUser.id,
    });
  });

  afterAll(async () => {
    await prismaService.user.delete({ where: { id: testUser.id } });
    await app.close();
  });

  it('should update a recipe', async () => {
    return request(app.getHttpServer())
      .patch('/recipe/1')
      .send(correctPayload)
      .set('Authorization', `bearer ${accessToken}`)
      .expect(HttpStatus.OK);
  });

  it('should throw error on unathourized user', async () => {
    return request(app.getHttpServer())
      .patch('/recipe/1')
      .send(correctPayload)
      .expect(HttpStatus.UNAUTHORIZED);
  });

  validationTestsStructure.map((object) => {
    const uncorrectPayload = { ...correctPayload };
    uncorrectPayload[object.property] = object.value;
    return it(`should throw validation error when property ${object.property} is ${object.value}`, async () => {
      return request(app.getHttpServer())
        .patch('/recipe/1')
        .send(uncorrectPayload)
        .set('Authorization', `bearer ${accessToken}`)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
