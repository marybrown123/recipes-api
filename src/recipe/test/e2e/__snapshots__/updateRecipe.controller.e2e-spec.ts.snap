// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe Controller - Update should throw validation error when property description is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "description must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`Recipe Controller - Update should throw validation error when property fileId is test 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "fileId must be a number conforming to the specified constraints",
  ],
  "statusCode": 400,
}
`;

exports[`Recipe Controller - Update should throw validation error when property ingredients is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "nested property ingredients must be either object or array",
  ],
  "statusCode": 400,
}
`;

exports[`Recipe Controller - Update should throw validation error when property name is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "name must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`Recipe Controller - Update should throw validation error when property preparing is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "nested property preparing must be either object or array",
  ],
  "statusCode": 400,
}
`;
